File: accudrop/app/src/main/java/me/chrislane/accudrop/network/CoordSender.java
package me.chrislane.accudrop.network;

import android.os.AsyncTask;
import android.os.Handler;
import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

public class CoordSender implements Runnable {

    private static final String TAG = CoordSender.class.getSimpleName();
    private final Socket socket;
    private final Handler handler;
    private OutputStream out;

    CoordSender(Socket socket, Handler handler) {
        this.socket = socket;
        this.handler = handler;
    }

    @Override
    public void run() {
        try {
            InputStream in = socket.getInputStream();
            out = socket.getOutputStream();
            byte[] buffer = new byte[1024];
            int bytes;
            handler.obtainMessage(Peer2Peer.COORD_SENDER, this).sendToTarget();

            while (true) {
                try {
                    bytes = in.read(buffer);
                    if (bytes == -1) {
                        break;
                    }

                    handler.obtainMessage(Peer2Peer.COORD_MSG, bytes, -1, buffer)
                            .sendToTarget();
                } catch (IOException e) {
                    Log.e(TAG, "Exception in loop: ", e);
                }
            }
        } catch (IOException e) {
            Log.e(TAG, "Exception in run: ", e);
        } finally {
            try {
                socket.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void write(byte[] buffer) {
        // TODO: Send data as anything but strings...
        AsyncTask.execute(() -> {
            try {
                out.write(buffer);
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }
}

File: app/src/main/java/me/chrislane/accudrop/network/CoordSender.kt
package me.chrislane.accudrop.network

import android.os.AsyncTask
import android.os.Handler
import android.util.Log
import java.io.IOException
import java.io.OutputStream
import java.net.Socket

class CoordSender internal constructor(private val socket: Socket?, private val handler: Handler) : Runnable {
    private var out: OutputStream? = null

    override fun run() {
        try {
            val `in` = socket!!.getInputStream()
            out = socket.getOutputStream()
            val buffer = ByteArray(1024)
            var bytes: Int
            handler.obtainMessage(Peer2Peer.COORD_SENDER, this).sendToTarget()

            while (true) {
                bytes = `in`.read(buffer)
                if (bytes == -1) {
                    break
                }
                handler.obtainMessage(Peer2Peer.COORD_MSG, bytes, -1, buffer)
                        .sendToTarget()
            }
        } catch (e: IOException) {
            Log.e(TAG, "Exception in run: ", e)
        } finally {
            if (socket != null && socket.isConnected) {
                try {
                    socket.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
    }

    fun write(buffer: ByteArray) {
        // TODO: Send data as anything but strings...
        AsyncTask.execute {
            try {
                out?.write(buffer)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private val TAG = CoordSender::class.java.simpleName
    }
}
